File: machine/cpp/machine/devices/vdp/readback.cpp
#include "machine/devices/vdp/readback.h"

#include "machine/bus/io.h"
#include "machine/cpu/cpu.h"
#include "machine/devices/device_status.h"
#include "machine/memory/memory.h"
#include <algorithm>

namespace bmsx {
VdpReadbackUnit::VdpReadbackUnit(Memory& memory, DeviceStatusLatch& fault)
	: m_memory(memory)
	, m_fault(fault) {
}

void VdpReadbackUnit::invalidateFrameBuffer() {
	m_frameBufferCache.width = 0u;
}

void VdpReadbackUnit::beginFrame() {
	m_readBudgetBytes = ReadbackBudgetBytes;
	m_readOverflow = false;
}

u32 VdpReadbackUnit::status() const {
	u32 value = 0u;
	if (m_readBudgetBytes >= 4u) {
		value |= VDP_RD_STATUS_READY;
	}
	if (m_readOverflow) {
		value |= VDP_RD_STATUS_OVERFLOW;
	}
	return value;
}

u32 VdpReadbackUnit::read(const VdpSurfaceBacking& surface, u32 requestedSurfaceId, u32 mode, u32 x, u32 y) {
	if (mode != VDP_RD_MODE_RGBA8888) {
		m_fault.raise(VDP_FAULT_RD_UNSUPPORTED_MODE, mode);
		return 0u;
	}
	if (requestedSurfaceId != VDP_RD_SURFACE_FRAMEBUFFER) {
		m_fault.raise(VDP_FAULT_RD_SURFACE, requestedSurfaceId);
		return 0u;
	}
	if (x >= surface.surfaceWidth || y >= surface.surfaceHeight) {
		m_fault.raise(VDP_FAULT_RD_OOB, x | (y << 16u));
		return 0u;
	}
	if (m_readBudgetBytes < 4u) {
		m_readOverflow = true;
		return 0u;
	}
	const ReadCache& cache = getReadCache(surface, x, y);
	const u32 localX = x - cache.x0;
	const size_t byteIndex = static_cast<size_t>(localX) * 4u;
	const u32 r = cache.data[byteIndex];
	const u32 g = cache.data[byteIndex + 1u];
	const u32 b = cache.data[byteIndex + 2u];
	const u32 a = cache.data[byteIndex + 3u];
	m_readBudgetBytes -= 4u;
	u32 nextX = x + 1u;
	u32 nextY = y;
	if (nextX >= surface.surfaceWidth) {
		nextX = 0u;
		nextY = y + 1u;
	}
	m_memory.writeValue(IO_VDP_RD_X, valueNumber(static_cast<double>(nextX)));
	m_memory.writeValue(IO_VDP_RD_Y, valueNumber(static_cast<double>(nextY)));
	return r | (g << 8u) | (b << 16u) | (a << 24u);
}

VdpReadbackState VdpReadbackUnit::captureState() const {
	VdpReadbackState state;
	state.readBudgetBytes = m_readBudgetBytes;
	state.readOverflow = m_readOverflow;
	return state;
}

void VdpReadbackUnit::restoreState(const VdpReadbackState& state) {
	m_readBudgetBytes = state.readBudgetBytes;
	m_readOverflow = state.readOverflow;
}

VdpReadbackUnit::ReadCache& VdpReadbackUnit::getReadCache(const VdpSurfaceBacking& surface, u32 x, u32 y) {
	ReadCache& cache = m_frameBufferCache;
	if (cache.width == 0u || cache.y != y || x < cache.x0 || x >= cache.x0 + cache.width) {
		prefetchReadCache(cache, surface, x, y);
	}
	return cache;
}

// start numeric-sanitization-acceptable -- readback chunk width is the minimum of hardware cap, remaining surface span, and per-frame read budget.
void VdpReadbackUnit::prefetchReadCache(ReadCache& cache, const VdpSurfaceBacking& surface, u32 x, u32 y) {
	const u32 maxPixelsByBudget = m_readBudgetBytes / 4u;
	if (maxPixelsByBudget == 0u) {
		m_readOverflow = true;
		cache.width = 0u;
		return;
	}
	const u32 chunkW = std::min(ReadbackMaxChunkPixels, std::min(surface.surfaceWidth - x, maxPixelsByBudget));
	copySurfacePixels(cache, surface, x, y, chunkW, 1u);
	cache.x0 = x;
	cache.y = y;
	cache.width = chunkW;
}
// end numeric-sanitization-acceptable

void VdpReadbackUnit::copySurfacePixels(ReadCache& cache, const VdpSurfaceBacking& surface, u32 x, u32 y, u32 width, u32 height) {
	const u32 stride = surface.surfaceWidth * 4u;
	const u32 rowBytes = width * 4u;
	std::array<u8, ReadbackMaxChunkPixels * 4u>& out = cache.data;
	for (u32 row = 0u; row < height; ++row) {
		const size_t srcOffset = static_cast<size_t>(y + row) * static_cast<size_t>(stride) + static_cast<size_t>(x) * 4u;
		const size_t dstOffset = static_cast<size_t>(row) * static_cast<size_t>(rowBytes);
		for (size_t byteIndex = 0u; byteIndex < rowBytes; ++byteIndex) {
			out[dstOffset + byteIndex] = surface.cpuReadback[srcOffset + byteIndex];
		}
	}
}

} // namespace bmsx

File: machine/ts/machine/devices/vdp/readback.ts
import {
	IO_VDP_RD_X,
	IO_VDP_RD_Y,
} from '../../bus/io';
import type { Memory } from '../../memory/memory';
import type { DeviceStatusLatch } from '../device_status';
import {
	VDP_FAULT_RD_OOB,
	VDP_FAULT_RD_SURFACE,
	VDP_FAULT_RD_UNSUPPORTED_MODE,
	VDP_RD_MODE_RGBA8888,
	VDP_RD_STATUS_OVERFLOW,
	VDP_RD_STATUS_READY,
	VDP_RD_SURFACE_FRAMEBUFFER,
} from './contracts';
import type { VdpSurfaceBacking } from './device_output';

export type VdpReadbackState = {
	readBudgetBytes: number;
	readOverflow: boolean;
};

type VdpReadCache = {
	x0: number;
	y: number;
	width: number;
	data: Uint8Array;
};

const VDP_READBACK_BUDGET_BYTES = 4096;
const VDP_READBACK_MAX_CHUNK_PIXELS = 256;

export class VdpReadbackUnit {
	private readonly frameBufferCache: VdpReadCache = { x0: 0, y: 0, width: 0, data: new Uint8Array(VDP_READBACK_MAX_CHUNK_PIXELS * 4) };
	private readBudgetBytes = VDP_READBACK_BUDGET_BYTES;
	private readOverflow = false;

	public constructor(
		private readonly memory: Memory,
		private readonly fault: DeviceStatusLatch,
	) {
	}

	public invalidateFrameBuffer(): void {
		this.frameBufferCache.width = 0;
	}

	public beginFrame(): void {
		this.readBudgetBytes = VDP_READBACK_BUDGET_BYTES;
		this.readOverflow = false;
	}

	public status(): number {
		let status = 0;
		if (this.readBudgetBytes >= 4) {
			status |= VDP_RD_STATUS_READY;
		}
		if (this.readOverflow) {
			status |= VDP_RD_STATUS_OVERFLOW;
		}
		return status;
	}

	public read(surface: VdpSurfaceBacking, requestedSurfaceId: number, mode: number, x: number, y: number): number {
		if (mode !== VDP_RD_MODE_RGBA8888) {
			this.fault.raise(VDP_FAULT_RD_UNSUPPORTED_MODE, mode);
			return 0;
		}
		if (requestedSurfaceId !== VDP_RD_SURFACE_FRAMEBUFFER) {
			this.fault.raise(VDP_FAULT_RD_SURFACE, requestedSurfaceId);
			return 0;
		}
		const width = surface.surfaceWidth;
		const height = surface.surfaceHeight;
		if (x >= width || y >= height) {
			this.fault.raise(VDP_FAULT_RD_OOB, (x | (y << 16)) >>> 0);
			return 0;
		}
		if (this.readBudgetBytes < 4) {
			this.readOverflow = true;
			return 0;
		}
		const cache = this.getReadCache(surface, x, y);
		const localX = x - cache.x0;
		const byteIndex = localX * 4;
		const r = cache.data[byteIndex]!;
		const g = cache.data[byteIndex + 1]!;
		const b = cache.data[byteIndex + 2]!;
		const a = cache.data[byteIndex + 3]!;
		this.readBudgetBytes -= 4;
		let nextX = x + 1;
		let nextY = y;
		if (nextX >= width) {
			nextX = 0;
			nextY = y + 1;
		}
		this.memory.writeValue(IO_VDP_RD_X, nextX);
		this.memory.writeValue(IO_VDP_RD_Y, nextY);
		return (r | (g << 8) | (b << 16) | (a << 24)) >>> 0;
	}

	public captureState(): VdpReadbackState {
		return {
			readBudgetBytes: this.readBudgetBytes,
			readOverflow: this.readOverflow,
		};
	}

	public restoreState(state: VdpReadbackState): void {
		this.readBudgetBytes = state.readBudgetBytes;
		this.readOverflow = state.readOverflow;
	}

	private getReadCache(surface: VdpSurfaceBacking, x: number, y: number): VdpReadCache {
		const cache = this.frameBufferCache;
		if (cache.width === 0 || cache.y !== y || x < cache.x0 || x >= cache.x0 + cache.width) {
			this.prefetchReadCache(cache, surface, x, y);
		}
		return cache;
	}

	private prefetchReadCache(cache: VdpReadCache, surface: VdpSurfaceBacking, x: number, y: number): void {
		const maxPixelsByBudget = this.readBudgetBytes >>> 2;
		if (maxPixelsByBudget <= 0) {
			this.readOverflow = true;
			cache.width = 0;
			return;
		}
		const remainingWidth = surface.surfaceWidth - x;
		const chunkLimit = VDP_READBACK_MAX_CHUNK_PIXELS < remainingWidth ? VDP_READBACK_MAX_CHUNK_PIXELS : remainingWidth;
		const chunkW = chunkLimit < maxPixelsByBudget ? chunkLimit : maxPixelsByBudget;
		this.copySurfacePixels(cache, surface, x, y, chunkW, 1);
		cache.x0 = x;
		cache.y = y;
		cache.width = chunkW;
	}

	private copySurfacePixels(cache: VdpReadCache, surface: VdpSurfaceBacking, x: number, y: number, width: number, height: number): void {
		const buffer = surface.cpuReadback;
		const stride = surface.surfaceWidth * 4;
		const rowBytes = width * 4;
		const out = cache.data;
		for (let row = 0; row < height; row += 1) {
			const srcOffset = (y + row) * stride + x * 4;
			const dstOffset = row * rowBytes;
			for (let byte = 0; byte < rowBytes; byte += 1) {
				out[dstOffset + byte] = buffer[srcOffset + byte]!;
			}
		}
	}
}
